### PR TITLE
Feature/forgot password

### PR DESCRIPTION
**Pull request by rishabh-vinsol**

### CODE REVIEW

File: app/views/user_mailer/reset_password.html.erb [hunk 1]
```diff
@@ -0,0 +1,2 @@
+<h1>Hi <%= @user.first_name %></h1>
+<p>Click on this <%= link_to 'link',  forgot_password_edit_url(id: @user.id, token: @user.reset_token) %> to change your password</p>
```
Comment (himanshumishra31) [inline app/views/user_mailer/reset_password.html.erb:2]: create this method on user to get forgot link url

File: app/views/user_mailer/confirmation.html.erb [hunk 2]
```diff
@@ -1,2 +1,2 @@
 <h1>Hi <%= @user.first_name %></h1>
-<p>Click on this <%= link_to 'link', "http://127.0.0.1:3000/verify_email/?token=#{@user.auth_token}&id=#{@user.id}" %> to verify you email</p>
+<p>Click on this <%= link_to 'link', verify_email_url(id: @user.id, token: @user.auth_token) %> to verify you email</p>
```
Comment (himanshumishra31) [inline app/views/user_mailer/confirmation.html.erb:2]: same as below

File: app/models/user.rb [hunk 3]
```diff
@@ -23,6 +23,14 @@ def update_verified_at
     update(verified_at: Time.now)
   end
 
+  def reset_token_valid?(token)
+    reset_token == token
+  end
+
+  def generate_reset_token
+    update(reset_token: SecureRandom.base64)
```
Comment (himanshumishra31) [inline app/models/user.rb:31]: create a RandomTokenGenerator to create unique token

Comment (himanshumishra31) [reply app/models/user.rb:31]: Add a service folder and add this service in that.

File: app/controllers/sessions_controller.rb [hunk 4]
```diff
@@ -1,7 +1,10 @@
 # Controller to handle users login
 class SessionsController < ApplicationController
+  layout 'user'
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/sessions_controller.rb:3]: fix this.

File: app/views/users/_form.html.erb [hunk 5]
```diff
@@ -1,32 +1,24 @@
 <%= form_with(model: user, url: signup_path, class:"mx-1 mx-md-4")  do |form| %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/users/_form.html.erb:1]: This form should not be submitted as a signup. It should be users create and get redirected to users index after creation.

File: app/views/users/_form.html.erb [hunk 6]
```diff
@@ -1,32 +1,24 @@
...
   </div>
-  <div class="d-flex flex-row align-items-center mb-4">
-    <i class="fas fa-user fa-lg me-3 fa-fw"></i>
-    <div class="form-outline flex-fill mb-0">
-      <%= form.text_field :last_name, class:"form-control" %>
-      <%= form.label :last_name, class:"form-label"%>
-    </div>
+  <div class="mb-3">
+    <%= form.text_field :last_name, class:"form-control" %>
+    <%= form.label :last_name, class:"form-label"%>
```
Comment (Sparsh-Goyal-2000) [inline app/views/users/_form.html.erb:9]: Have you removed the role field ?

**Comment by Sparsh-Goyal-2000** [comment]
@rishabh-vinsol 
Why have you created so many controllers just for login functionality ?
There should be just a user controller and session controller.
You can move actions related to user (verify_email, forgot and change_password) to users_controller.
Check if we can use users create for signup action as well.

File: app/helpers/store_helper.rb [hunk 7]
```diff
@@ -0,0 +1,2 @@
+module StoreHelper
```
Comment (Sparsh-Goyal-2000) [inline app/helpers/store_helper.rb:1]: remove this for now.

File: app/models/user.rb [hunk 8]
```diff
@@ -23,6 +23,14 @@ def update_verified_at
     update(verified_at: Time.now)
```
Comment (Sparsh-Goyal-2000) [inline app/models/user.rb:23]: update method name like `verified`

File: app/controllers/application_controller.rb [hunk 9]
```diff
@@ -1,2 +1,19 @@
 class ApplicationController < ActionController::Base
+  before_action :authorize
+
+  private def redirect_to_homepage_if_logged_in
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/application_controller.rb:4]: Update method name

File: app/controllers/application_controller.rb [hunk 10]
```diff
@@ -1,2 +1,19 @@
...
+    return unless session[:user_id]
+
+    redirect_to store_index_path, notice: t('notice.application.logged_in')
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/application_controller.rb:7]: user root path here, and update that in routes file if required.

File: app/controllers/application_controller.rb [hunk 11]
```diff
@@ -1,2 +1,19 @@
...
+  end
+
+  private def authorize
+    return if current_user
+
+    redirect_to login_url, alert: t('notice.application.login_request')
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/application_controller.rb:13]: you have used login_url and login_path.
Is there any difference?
If not, use same method everywhere.

File: app/controllers/confirmations_controller.rb [hunk 12]
```diff
@@ -1,5 +1,7 @@
 # Controller for user email verification after signup
 class ConfirmationsController < ApplicationController
+  layout 'user'
+  skip_before_action :authorize
   before_action :set_user
 
   def verify_email
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/confirmations_controller.rb:7]: You can move this in users controller.

File: app/controllers/store_controller.rb [hunk 13]
```diff
@@ -0,0 +1,2 @@
+class StoreController < ApplicationController
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/store_controller.rb:1]: We use plural names for controllers.

File: app/models/concerns/authenticatable.rb [hunk 14]
```diff
@@ -8,4 +8,8 @@ module Authenticatable
   def send_authentication_email
```
Comment (Sparsh-Goyal-2000) [inline app/models/concerns/authenticatable.rb:8]: no need to create concern if using in single file
Check if we can move these methods in user.rb

File: app/views/layouts/_navbar.html.erb [hunk 15]
```diff
@@ -15,17 +23,17 @@
...
-        <input class="form-control me-2" type="search" placeholder="Search" aria-label="Search">
-        <button class="btn btn-outline-light" type="submit">Search</button>
-      </form>
-      <button type="button" class="btn btn-outline-light">
-        <i class="bi bi-cart-fill"></i>
-        Cart
-      </button>
+      <div class="d-flex align-items-center">
+      <% if session[:user_id] %> 
+        <%= button_to "Logout", logout_path, method: :delete, class: "btn btn-outline-secondary me-3" %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/layouts/_navbar.html.erb:31]: Show here logged_in as `user name`.
we can add a button to users profile as well.

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]

**Comment by Sparsh-Goyal-2000** [comment]
@rishabh-vinsol merge main in this branch

File: app/views/user_mailer/reset_password.html.erb (hunk 1 above)
Comment (rishabh-vinsol) [reply app/views/user_mailer/reset_password.html.erb:2]: Done

File: app/views/user_mailer/confirmation.html.erb (hunk 2 above)
Comment (rishabh-vinsol) [reply app/views/user_mailer/confirmation.html.erb:2]: Done

File: app/views/layouts/user.html.erb [hunk 16]
```diff
@@ -0,0 +1,36 @@
...
+  <head>
+    <title>Champs</title>
+    <meta name="viewport" content="width=device-width,initial-scale=1">
+    <%= csrf_meta_tags %>
+    <%= csp_meta_tag %>
+    <%= stylesheet_link_tag "application", "data-turbo-track": "reload" %>
+    <%= javascript_include_tag 'show_password' %>
+    <%= javascript_importmap_tags %>
+  </head>
+  <body>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/layouts/user.html.erb:12]: Done

File: app/controllers/confirmations_controller.rb [hunk 17]
```diff
@@ -1,5 +1,7 @@
 # Controller for user email verification after signup
 class ConfirmationsController < ApplicationController
+  layout 'user'
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/confirmations_controller.rb:3]: DOne

File: app/controllers/passwords_controller.rb [hunk 18]
```diff
@@ -0,0 +1,45 @@
+class PasswordsController < ApplicationController
+  layout "user"
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/passwords_controller.rb:2]: Done

File: app/controllers/registrations_controller.rb [hunk 19]
```diff
@@ -1,7 +1,10 @@
 # Controller to signup new users
 class RegistrationsController < ApplicationController
+  layout 'user'
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/registrations_controller.rb:3]: Done

File: app/controllers/sessions_controller.rb (hunk 4 above)
Comment (rishabh-vinsol) [reply app/controllers/sessions_controller.rb:3]: Done

File: app/helpers/store_helper.rb (hunk 7 above)
Comment (rishabh-vinsol) [reply app/helpers/store_helper.rb:1]: Done

File: app/models/user.rb (hunk 8 above)
Comment (rishabh-vinsol) [reply app/models/user.rb:23]: Done

File: app/controllers/application_controller.rb (hunk 10 above)
Comment (rishabh-vinsol) [reply app/controllers/application_controller.rb:7]: Done

File: app/controllers/confirmations_controller.rb (hunk 12 above)
Comment (rishabh-vinsol) [reply app/controllers/confirmations_controller.rb:7]: Done

File: app/controllers/store_controller.rb (hunk 13 above)
Comment (rishabh-vinsol) [reply app/controllers/store_controller.rb:1]: Removed this controller for now

File: app/models/concerns/authenticatable.rb (hunk 14 above)
Comment (rishabh-vinsol) [reply app/models/concerns/authenticatable.rb:8]: Done

File: app/views/layouts/_navbar.html.erb (hunk 15 above)
Comment (rishabh-vinsol) [reply app/views/layouts/_navbar.html.erb:31]: Sir will do this in another PR.

File: app/models/user.rb [hunk 20]
```diff
@@ -1,33 +1,62 @@
...
+  def generate_reset_token
+    update(reset_token: RandomTokenGenerator.new('User', 'auth_token').generate_token)
+  end
+
+  def send_authentication_email
+    UserMailer.with(user_id: id).confirmation.deliver_now
+  end
+
+  def send_password_reset_email
+    UserMailer.with(user_id: id).reset_password.deliver_now
```
Comment (Sparsh-Goyal-2000) [inline app/models/user.rb:47]: same

File: app/controllers/registrations_controller.rb [hunk 21]
```diff
@@ -1,24 +1,107 @@
...
+    end
+  end
+
+  def login_user
+    if @user && !@user.verified_at?
+      flash.now[:alert] = t('.not_verified')
+      render :login
+    elsif @user.try(:authenticate, params[:password])
+      session[:user_id] = @user.id
+      redirect_to "/"
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/registrations_controller.rb:31]: can we redirect to root path ?

File: app/views/users/_form.html.erb [hunk 22]
```diff
@@ -1,32 +1,28 @@
...
+  <div class="mb-3">
+    <%= form.label :password, class:"form-label"%>
+    <%= form.password_field :password, class:"form-control" %>
+  </div>
+  <div class="d-flex flex-row align-items-center mb-5">
+    <%= form.check_box :show_password, onchange: "showPassword()", class: "form-check-input me-2" %>
+    <%= form.label :show_password, 'Show password', class: 'form-check-label' %>
   </div>
   <div class="d-flex justify-content-center mx-4 mb-3 mb-lg-4">
     <%= form.submit class:"btn btn-primary btn-lg" %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/users/_form.html.erb:28]: @rishabh-vinsol 
in users edit view
pls remove show this user link, as you have removed show route
Also `Back to users` is not visible below
Add that at the top where you are showing edit user or new user
like -> `Users/New User` -> here `User` will be url for users index page

Make this change wherever the links are not visible.

File: config/routes.rb [hunk 23]
```diff
@@ -2,21 +2,19 @@
...
     get :verify_email
-  end
-
-  controller :sessions do
-    get 'login' => :new
-    post 'login' => :create
-    delete 'logout' => :destroy
+    get :forgot_password
+    post 'forgot_password' => :forgot_password_email
+    get 'forgot_password/edit' => :forgot_password_edit
```
Comment (Sparsh-Goyal-2000) [inline config/routes.rb:16]: rename the url as reset password.

File: app/models/user.rb [hunk 24]
```diff
@@ -1,38 +1,67 @@
...
+  end
+
+  def email_verification_url(host)
+    Rails.application.routes.url_helpers.verify_email_url(id: id, token: auth_token, host: host)
+  end
+
+  def reset_password_url(host)
+    Rails.application.routes.url_helpers.forgot_password_edit_url(id: id, token: reset_token, host: host)
```
Comment (Sparsh-Goyal-2000) [inline app/models/user.rb:55]: same

File: app/controllers/registrations_controller.rb [hunk 25]
```diff
@@ -1,24 +1,107 @@
...
+      @user.generate_reset_token
+      @user.send_password_reset_email
+      redirect_to login_path, notice: t('.reset_password_email')
+    else
+      flash.now[:alert] = t('.invalid_user', email: params[:email])
+      render :forgot_password
+    end
+  end
+
+  def forgot_password_edit
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/registrations_controller.rb:66]: reset_password

File: app/controllers/registrations_controller.rb [hunk 26]
```diff
@@ -1,24 +1,107 @@
...
+    if @user.reset_token_valid?(params[:token])
+      render :forgot_password_edit
+    else
+      redirect_to login_path, alert: t('.auth_unsuccessful')
+    end
+  end
+
+  def update_password
+    @user = User.find(user_params[:id])
+    if @user.update(user_params)
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/registrations_controller.rb:76]: only allow password here

File: app/models/user.rb [hunk 27]
```diff
@@ -1,38 +1,67 @@
...
+  }
 
   ### VALIDATIONS ###
 
   validates :first_name, :last_name, :email, presence: true
   validates :email, uniqueness: true, email: true
 
+  ### CALLBACKS ###
+
+  before_create :generate_auth_token
```
Comment (Sparsh-Goyal-2000) [inline app/models/user.rb:16]: it should be like set_auth_token as it is also setting the value on user

File: app/services/random_token_generator.rb [hunk 28]
```diff
@@ -0,0 +1,14 @@
+class RandomTokenGenerator
+  def initialize(model_name, model_token)
+    @model_name = model_name.constantize
+    @model_token = model_token
```
Comment (Sparsh-Goyal-2000) [inline app/services/random_token_generator.rb:4]: rename this to token_name
as one model can have 2 token like user.

File: app/views/registrations/forgot_password.html.erb [hunk 29]
```diff
@@ -0,0 +1,13 @@
...
+<%= form_with do |form| %>
+  <div class="mb-3">
+    <%= render 'shared/notice' %>
+  </div>
+  <div class="mb-3">
+    <%= form.label :email, "Enter the email", class:"form-label" %>
+    <%= form.text_field :email, class:"form-control"%>
+  </div>
+  <div class="d-flex justify-content-center mx-4 mb-3 mb-lg-4">
+    <%= form.submit "Reset", class:"btn btn-primary btn-lg" %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/registrations/forgot_password.html.erb:11]: Label should be `reset password`

File: config/environments/development.rb [hunk 30]
```diff
@@ -79,4 +82,4 @@
 
   # Uncomment if you wish to allow Action Cable access from any origin.
   # config.action_cable.disable_request_forgery_protection = true
-end
+end
```
Comment (Sparsh-Goyal-2000) [inline config/environments/development.rb:85]: add endline

File: app/views/registrations/login.html.erb [hunk 31]
```diff
@@ -0,0 +1,29 @@
+<p class="text-center h1 fw-bold mb-5 mx-1 mx-md-4 mt-4">Login</p>
```
Comment (Sparsh-Goyal-2000) [inline app/views/registrations/login.html.erb:1]: @rishabh-vinsol 
currently if a user signup but deleted the verify account email by mistake.
How can he login or get the mail again?

File: app/controllers/users_controller.rb [hunk 32]
```diff
@@ -13,28 +13,28 @@ def create
     @user = User.new(user_params)
```
Comment (Sparsh-Goyal-2000) [inline app/controllers/users_controller.rb:13]: on the user listing page
`def index
    @users = User.all
  end`
how are these users sorted?
Sort them with id desc

**Review by Sparsh-Goyal-2000 — CHANGES_REQUESTED** [review]

File: config/locales/en.yml [hunk 33]
```diff
@@ -56,3 +69,16 @@ en:
   user_mailer:
     confirmation:
       subject: "OTP for email authentication"
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply config/locales/en.yml:71]: Done

File: config/locales/en.yml [hunk 34]
```diff
@@ -56,3 +69,16 @@ en:
...
+    reset_password:
+      subject: "OTP to reset password"
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply config/locales/en.yml:73]: Done

File: app/views/user_mailer/confirmation.html.erb [hunk 35]
```diff
@@ -1,2 +1,2 @@
 <h1>Hi <%= @user.first_name %></h1>
-<p>Click on this <%= link_to 'link', "http://127.0.0.1:3000/verify_email/?token=#{@user.auth_token}&id=#{@user.id}" %> to verify you email</p>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/user_mailer/confirmation.html.erb:2]: Done

File: app/views/user_mailer/reset_password.html.erb [hunk 36]
```diff
@@ -0,0 +1,2 @@
+<h1>Hi <%= @user.first_name %></h1>
+<p>Click on this <%= link_to 'link',  @user.reset_password_url(@host) %> to change your password</p>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/user_mailer/reset_password.html.erb:2]: Done

File: app/views/users/_form.html.erb [hunk 37]
```diff
@@ -1,32 +1,28 @@
...
+    <%= form.label :email, class:"form-label"%>
+    <%= form.email_field :email, class:"form-control" %>
   </div>
-  <div class="d-flex flex-row align-items-center mb-4">
-    <i class="fas fa-user fa-lg me-3 fa-fw"></i>
-    <div class="form-outline flex-fill mb-0">
-      <%= form.password_field :password, class:"form-control" %>
-      <%= form.label :password, class:"form-label"%>
-    </div>
+  <div class="mb-3">
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/users/_form.html.erb:15]: Done

File: app/views/registrations/signup.html.erb [hunk 38]
```diff
@@ -0,0 +1,2 @@
+<p class="text-center h1 fw-bold mb-5 mx-1 mx-md-4 mt-4">Sign up</p>
+<%= render "users/form", user: @user %>
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/views/registrations/signup.html.erb]: Did the changes

File: app/controllers/registrations_controller.rb [hunk 39]
```diff
@@ -1,24 +1,107 @@
...
+    @user = User.find_by(email: params[:email])
+  end
+
   private def user_params
-    params.require(:user).permit(:first_name, :last_name, :email, :password)
+    params.require(:user).permit(:id, :first_name, :last_name, :email, :password, :password_confirmation)
+  end
+
+  private def invalid_user
+    redirect_to forgot_password_path, alert: t('errors.passwords.invalid_user')
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/registrations_controller.rb:96]: It should be redirected to login_path

File: app/models/user.rb [hunk 40]
```diff
@@ -1,33 +1,62 @@
...
+  def reset_token_valid?(token)
+    reset_token == token
+  end
+
+  def generate_reset_token
+    update(reset_token: RandomTokenGenerator.new('User', 'auth_token').generate_token)
+  end
+
+  def send_authentication_email
+    UserMailer.with(user_id: id).confirmation.deliver_now
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/models/user.rb:43]: Done

File: app/models/user.rb (hunk 20 above)
Comment (rishabh-vinsol) [reply app/models/user.rb:47]: Done

File: app/controllers/registrations_controller.rb (hunk 21 above)
Comment (rishabh-vinsol) [reply app/controllers/registrations_controller.rb:31]: Yes sir, Done

File: app/views/users/_form.html.erb (hunk 22 above)
Comment (rishabh-vinsol) [reply app/views/users/_form.html.erb:28]: Done

File: app/models/user.rb [hunk 41]
```diff
@@ -1,38 +1,67 @@
...
+  end
+
+  def send_password_reset_email
+    UserMailer.with(user_id: id).reset_password.deliver_now
+  end
+
+  def email_verification_url(host)
+    Rails.application.routes.url_helpers.verify_email_url(id: id, token: auth_token, host: host)
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/models/user.rb:51]: Done

File: app/views/registrations/forgot_password.html.erb (hunk 29 above)
Comment (rishabh-vinsol) [reply app/views/registrations/forgot_password.html.erb:11]: Done

File: config/routes.rb (hunk 23 above)
Comment (rishabh-vinsol) [reply config/routes.rb:16]: Done

File: app/models/user.rb (hunk 24 above)
Comment (rishabh-vinsol) [reply app/models/user.rb:55]: Done

File: app/controllers/registrations_controller.rb (hunk 25 above)
Comment (rishabh-vinsol) [reply app/controllers/registrations_controller.rb:66]: Done

File: app/models/user.rb (hunk 27 above)
Comment (rishabh-vinsol) [reply app/models/user.rb:16]: Done

File: app/services/random_token_generator.rb (hunk 28 above)
Comment (rishabh-vinsol) [reply app/services/random_token_generator.rb:4]: Done

File: config/environments/development.rb (hunk 30 above)
Comment (rishabh-vinsol) [reply config/environments/development.rb:85]: Done

File: app/controllers/users_controller.rb (hunk 32 above)
Comment (rishabh-vinsol) [reply app/controllers/users_controller.rb:13]: Done

File: app/controllers/registrations_controller.rb [hunk 42]
```diff
@@ -1,24 +1,107 @@
...
+  end
+
+  private def token_presence
+    return unless params[:token]
```
[the remark this answers is not in the document]
Comment (rishabh-vinsol) [reply app/controllers/registrations_controller.rb:100]: Removed this.. Not needed

File: app/controllers/registrations_controller.rb (hunk 26 above)
Comment (rishabh-vinsol) [reply app/controllers/registrations_controller.rb:76]: Done

File: app/views/layouts/_navbar.html.erb [hunk 43]
```diff
@@ -1,31 +1,39 @@
...
-      <span class="navbar-toggler-icon"></span>
+    <button
+      class="navbar-toggler"
+      type="button"
+      data-mdb-toggle="collapse"
+      data-mdb-target="#navbarButtonsExample"
+      aria-controls="navbarButtonsExample"
+      aria-expanded="false"
+      aria-label="Toggle navigation"
+    >
```
Comment (Sparsh-Goyal-2000) [inline app/views/layouts/_navbar.html.erb:14]: This is not looking good
move `class="navbar-toggler"` after buttton.
then fix alignment and move `>` after area-label.

File: app/views/users/_form.html.erb [hunk 44]
```diff
@@ -1,32 +1,28 @@
...
+    <%= form.label :role, class:"form-label"%>
+    <%= form.select :role, User.roles.to_a.map {|r| [r[0].capitalize, r[0]]}, {}, {class: "form-select"} %>
+  </div>
+  <div class="mb-3">
+    <%= form.label :password, class:"form-label"%>
+    <%= form.password_field :password, class:"form-control" %>
```
Comment (Sparsh-Goyal-2000) [inline app/views/users/_form.html.erb:21]: Add password confirmation field.

**Review by Sparsh-Goyal-2000 — APPROVED** [review]

**Pull request merged**